### PR TITLE
[RFC] Multiple Cabals Continuation

### PR DESCRIPTION
**Pull request by cblgh**
this is a continuation of nick's pr #84 

i fucked up my branches so here we are with a new pr

### CODE REVIEW

File: commands.js [hunk 1]
```diff
@@ -71,6 +71,10 @@ function Commander (view, cabal) {
           self.view.writeLine.bind(self.view)(`/${key}`)
           self.view.writeLine.bind(self.view)(`  ${self.commands[key].help()}`)
         }
+        self.view.writeLine.bind(self.view)(`ctrl+w`)
```
Comment (nikolaiwarner) [inline commands.js]: ctrl+j ?

Comment (cblgh) [reply commands.js]: thanks! updated to `alt-n` (which is the new one lol)

**Comment by cblgh** [comment]
### TODO
- [ ] implement `/add <cabal`
- [ ] improve dx by shortening `self.state.cabal.client` (without breaking stuff this time)
- [ ] debug why `publishSingleMessage` doesn't work anymore
- [ ] publish cabal-core 3.1.0 

(feel free to edit and add more)

i want to get this into master so that we don't have anymore nasty merge conflicts from nice contribs <3

**Comment by nikolaiwarner** [comment]
@cblgh I tested it out and `publishSingleMessage` is working for me as is

other news:
- the `alt-n` command doesnt work for me as it want to print ˜ instead of switching panes
- i have the /add command in progress at around 25% complete

**Comment by cblgh** [comment]
hell YEAH we've got `/add <cabal` as a command now!

some issues i noticed from testing:
- [ ] lost names of everyone in previously joined cabals
- [ ]  created a channel with the name of my peer's public hash
- [ ]  lost names in joined cabal
- [ ]  names panel isn't updated

File: cli.js [hunk 2]
```diff
@@ -29,57 +37,125 @@ var usage = `Usage
...
-  cabal.db.ready(function () {
-    start(args.key)
-  })
-} else {
-  cabal = Cabal(args.db, null)
-  cabal.db.ready(function () {
-    cabal.getLocalKey(function (err, key) {
-      if (err) throw err
-      start(key)
+if (args.new) {
```
[the remark this answers is not in the document]
Comment (cblgh) [reply cli.js]: ohhhhhhhhhh this is exactly what i wanted to do, i just didn't know where to look! thanks noffle, it's fixed in the latest push :)

File: cli.js [hunk 3]
```diff
@@ -2,25 +2,33 @@
...
+var fs = require('fs')
+var yaml = require('js-yaml')
+var mkdirp = require('mkdirp')
 var frontend = require('./neat-screen.js')
 
 var args = minimist(process.argv.slice(2))
 
-var homedir = process.env.HOME || process.env.HOMEPATH || process.env.USERPROFILE
-var rootdir = args.dir || (homedir + '/.cabal/archives/')
+var protocolMajorVersion = Cabal.protocolVersion.split('.')[0]
```
[the remark this answers is not in the document]
Comment (cblgh) [reply cli.js]: no, that only happens if we make a breaking change (major version). and if we make a breaking change, i think we should be free to rid ourselves of the headache of old configs and make the best of the breaking change.

this is definitely to try to pin down our issues. we _should_ never need to do this, but _if_ we do it then we can exclude for certain that old clients are screwing things up.

i am fine with trying out this approach for a while (after this pr is stable and merged) and then reverting to not using a protocol version if it still feels really bad

File: cli.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (cblgh) [reply cli.js]: btw thanks for the review noffle, it's really nice to have feedback like this <3

**Comment by noffle** [comment]
Let's call it "database version" or something then, since it's only being used to make storage decisions.

File: cli.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (noffle) [reply cli.js]: I think we're saying the same things:

- I agree with this change (using a new dir to escape the vestigial hypercore issue)
- I don't like it being tied to *protocol version* though, since this is more like a *database version* bump
- If we conflate protocol and db versioning, then every time we do a major protocol change we'll be essentially deleting (hiding) users' private keys needlessly by creating a new cabal directory

File: cli.js [hunk 4]
```diff
@@ -29,57 +38,109 @@ var usage = `Usage
...
 if (args.key) {
-  args.key = args.key.replace('cabal://', '').replace('cbl://', '').replace('dat://', '').replace(/\//g, '')
-  args.db = rootdir + args.key
+  // If a key is provided, place it at the top of the list
+  cabalKeys.unshift(args.key)
+}
 
-  var cabal = Cabal(args.db, args.key)
+if (args.new) {
+  var key = crypto.keyPair().publicKey.toString('hex')
```
Comment (noffle) [inline cli.js]: :ok_hand:

File: package.json [hunk 5]
```diff
@@ -1,17 +1,20 @@
 {
   "name": "cabal",
-  "version": "5.0.0",
+  "version": "5.0.1",
```
Comment (noffle) [inline package.json]: Can we do version bump in a separate PR?

**Comment by cblgh** [comment]
after some discussions in #cabal-club we decided that it's better to get this into master and keep the show going

we still have bugs with multi-cabal, but the config stuff is really nice, so i'm going to try to put multi-cabal behind an `--experimental` flag and fix the bugs after the fact. i don't think affect single-cabal usage

### tasks 
- [ ] **_create_** `--experimental` cli flag to enable multi-cabal
- [ ] **_add_** alias section to config so people can do cabal --key cabal://00...ff --alias malmo-ppl; cabal --join malmo-ppl)
- [ ]  ***revert*** version bump and do it after pr has been merged into master
- [x] _**rename**_ protocolVersion to databaseVersion 
- [ ] ***merge*** into master
- [ ]  ***bump*** master version
- [ ] ***publish*** new cabal-core to npm
- [ ] ***publish*** cli to npm

File: neat-screen.js [hunk 6]
```diff
@@ -84,165 +95,246 @@ function NeatScreen (cabal) {
     else return
     this.bus.emit('render')
   })
+
+  // move between window panes with ctrl+j
+  this.neat.input.on('alt-n', () => {
```
Comment (nikolaiwarner) [inline neat-screen.js]: alt-n still doesnt work for me on mac terminal ... but we can solve it after this lands

**Comment by cblgh** [comment]
> @noffle: I noticed you're re-introducing package-lock.json -- intentional? LGTM otherwise!

that lil sucker just decided to appear, not intended!
